feat: Task Comments (UI + Backend)

yugantarjain opened the pull request:
### Description
This PR implements the task comments feature including fetching of task comments and posting a new comment on a task.
The PR also includes UI change to mark a task as complete, which is now done through a context menu (force/long press).

Note: the post new task comment API has some bugs in it and the GitHub issue for that exists. It inconsistently works for only some tasks and fails for others. However, this is not a blocker, since it does works in normal cases.

Note 2: the KeyboardAware() modifier that is used to lift the text field to the height of the keyboard isn't working perfectly here (some extra space is there). The modifier works perfectly in other areas of our app, this gives me the impression of a SwiftUI bug and not the code of the modifier. Also, fun fact: with Xcode 12, SwiftUI will handle this automatically and no custom code/modifier will be required!

Fixes #104 

### Mocks
<img width="498" alt="Screen Shot 2020-07-31 at 1 51 22 AM" src="https://user-images.githubusercontent.com/31164725/88971152-2345db80-d2d1-11ea-8118-59fe085a5d00.png">

### Type of Change:
- Code
- Quality Assurance
- User Interface
- Outreach
- Documentation

**Code/Quality Assurance Only**
- New feature (non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?
Tested on Simulator and iPhone 7.

### Checklist:
- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] Any dependent changes have been merged 

**Code/Quality Assurance Only**
- [x] My changes generate no new warnings 
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been published in downstream modules

yugantarjain committed:
context menu used for mark as complete and navigation link for comments in relation tasks to do.

yugantarjain committed:
tasks to do section converted to tasks section and tasks done section deleted. Tasks section used for both - to do and done.

yugantarjain committed:
tasks section updated to show correct icons for to do and done tasks.

yugantarjain committed:
navigation to task comments screen done.

yugantarjain committed:
Merge branch 'main'

Conflicts:
	mentorship ios.xcodeproj/project.pbxproj

yugantarjain committed:
Merge branch 'comments'

Conflicts:
	mentorship ios.xcodeproj/project.pbxproj

yugantarjain committed:
task time date formatter added. get task comments api UI + backend api implemented.

yugantarjain committed:
loading of task comments improved.

yugantarjain committed:
fetching of task comments perfected (ui + backend).

yugantarjain committed:
post a new task comment done. fetch task comments improved.

yugantarjain committed:
comments task code clean up.

yugantarjain committed:
Task Comments Tests added.

ApheleiaS reviewed (commented):
DISCLAIMER! These comments are related to the screen design and subject to discussion among mentors and Yugantar. 😄 

1. Would like more Padding in the list items 
2. The Commenter(text size) should be bigger than time/date info. Possibly in a different line above the date/time
3. Ordering: The newest comment should be on the top and the oldest at the bottom.

ApheleiaS commented on mentorship ios/ViewModels/TaskCommentsViewModel.swift:
```diff
@@ -0,0 +1,28 @@
+    
+    func getCommentAuthorName(authorID: Int) -> String {
+        if authorID == userID {
+            return "You"
```
localize this string, use a constant

yugantarjain replied:
Done

yugantarjain committed:
task comment list cell updated. padding added, fonts and font colors optimized.

yugantarjain committed:
Task name now shown in task comments screen. Show earlier functionality added.

yugantarjain committed:
localizable string constants updated.

yugantarjain commented:
@anitab-org/mentorship-ios-maintainers PR updated and all the changes have been made. It looks so much better now!

yugantarjain committed:
unit tests for task comments updated and complete.

yugantarjain committed:
swiftlint error resolved.

vatsalkul commented:
@yugantarjain I am not able to test this on my device! Is it a bug or something else. Here is a screen recording [Video](https://drive.google.com/file/d/135gfG7ZJ5L26rw1hHI4WydT9G0KxQzZv/view?usp=sharing)

yugantarjain commented:
> @yugantarjain I am not able to test this on my device! Is it a bug or something else. Here is a screen recording [Video](https://drive.google.com/file/d/135gfG7ZJ5L26rw1hHI4WydT9G0KxQzZv/view?usp=sharing)

This is probably being caused due to the bug in the post task comment API ([issue](https://github.com/anitab-org/mentorship-backend/issues/661) here). I should probably add an alert for such cases?

yugantarjain committed:
Alert added to notify the user when posting of comment fails.

yugantarjain commented:
@vatsalkul alert added to notify when posting of task comment is unsuccessful.

vatsalkul commented:
> @vatsalkul alert added to notify when posting of task comment is unsuccessful.

@yugantarjain do you know why is it happening exactly, Is there a possibility that you can solve the issue you opened in mentorship backend?

yugantarjain commented:
> > @vatsalkul alert added to notify when posting of task comment is unsuccessful.
> 
> @yugantarjain do you know why is it happening exactly, Is there a possibility that you can solve the issue you opened in mentorship backend?

I don't know about the exact bug in the API that is causing this. It is possible but I don't think it'll be feasible for me to work on the API bug.

sunjunkie commented:
1. Use more descriptive text in the TextField - e.g., "Enter Comment".  And also localize this.
2. I am seeing both task section titles are "Tasks To Do". This occurs on both the Tasks screen and the Dashboard. The bottom section should be "Tasks Done".

BTW, I am able to run this on my device. And I also have a task that comments cannot be added to (backend bug).

yugantarjain committed:
task to and done header fixed.

yugantarjain committed:
taks comment text field title made "Enter comment" and localized.

yugantarjain commented:
> 1. Use more descriptive text in the TextField - e.g., "Enter Comment".  And also localize this.
> 2. I am seeing both task section titles are "Tasks To Do". This occurs on both the Tasks screen and the Dashboard. The bottom section should be "Tasks Done".
> 
> BTW, I am able to run this on my device. And I also have a task that comments cannot be added to (backend bug).

@sunjunkie done.

yugantarjain committed:
conflicts resolved

sunjunkie commented:
I'm seeing some inconsistency in the comments (commentor).

**User sj:**
![Screen Shot 2020-08-04 at 9 51 49 AM](https://user-images.githubusercontent.com/399376/89322299-e4998380-d638-11ea-847e-42b839cf115b.png)


**User CG:**
![89322537-33471d80-d639-11ea-98a5-c2afd43f6f47](https://user-images.githubusercontent.com/399376/89322598-4a860b00-d639-11ea-85cf-3096e425018b.png)

ApheleiaS commented on mentorship iosTests/TaskCommentsTests.swift:
```diff
@@ -0,0 +1,142 @@
+        let taskCommentsService: TaskCommentsService = TaskCommentsAPI(urlSession: urlSession)
+        
+        // Set mock json and data
+        let mockJSON: [TaskCommentsModel.TaskCommentsResponse] = [
```
nit: mockJSON data for every test could go into a separate JSON file, and you could read that file in different tests to reusue the same data. Since most of the data would be related to the different Models we have, reuse is definitely possible.

yugantarjain replied:
Okay. Would doing that as a JSON file instead of using objects of the model have any advantage, or can we do it anyways? I also have another question, will this be a good approach for cases when we want to have specific data tailored to the test, and change that data within the test to cover different cases?

ApheleiaS commented on mentorship ios/Views/Tasks/TasksSection.swift:
```diff
@@ -0,0 +1,67 @@
+            //context menu used to show and enable actions (eg. mark as complete)
+            .contextMenu {
+                if self.isToDoSection && navToTaskComments {
+                    Button("Mark as Complete") { self.markAsCompleteAction(task) }
```
nit: Localizable String

yugantarjain replied:
Done

ApheleiaS commented on mentorship ios/ViewModels/TaskCommentsViewModel.swift:
```diff
@@ -0,0 +1,44 @@
+    @Published var taskCommentsResponse = [TaskCommentsModel.TaskCommentsResponse]()
+    @Published var newComment = TaskCommentsModel.PostCommentUploadData(comment: "")
+    @Published var postCommentResponse = TaskCommentsModel.PostCommentResponse(message: "", success: false)
+    @Published var isLoading = false
+    @Published var showingEarlier = false
+    @Published var showErrorAlert = false
```
nit: Should add comments about what these variables are for esp. `showingEarlier`

yugantarjain replied:
Done

ApheleiaS commented on mentorship ios/Views/Tasks/TaskComments.swift:
```diff
@@ -0,0 +1,129 @@
+            // if post comment successful, update comments and clear text field
+            if resp.success {
+                self.fetchComments()
+                self.taskCommentsVM.newComment.comment = ""
```
question: should we clear self.taskCommentsVM.postCommentResponse also here?

yugantarjain replied:
Not sure about this point, clear in the sense?

yugantarjain committed:
sorting logic for comments added. selection variable configured to show home page after login. inconsistent task comment authors name fixed.

yugantarjain committed:
unit test updated.

yugantarjain commented:
@anitab-org/mentorship-ios-maintainers 

I have fixed the inconsistency in commentor name issue. I probably wasn't very thorough last time, the backend returns the ids consistently, bug was in our app (the userID used in the view model remained of the previous user after logging in with a new user).

With this update, I've also fixed the bug where the settings screen was shown after logging out and logging back in the app, now the home screen is shown.

And lastly, sorting logic has been added to task comments array on the basis of their creation date.

yugantarjain committed:
mark as complete localizable constant used.

yugantarjain committed:
documentation improved.

yugantarjain committed:
Merge branch 'main'

Conflicts:
	mentorship ios.xcodeproj/project.pbxproj

yugantarjain committed:
Merge branch 'main'

Conflicts:
	mentorship ios.xcodeproj/project.pbxproj

yugantarjain committed:
Merge branch 'comments'

Conflicts:
	mentorship ios.xcodeproj/project.pbxproj

ApheleiaS reviewed (approved):
LGTM!

vatsalkul reviewed (approved):
Great work 🎉
